java-frontend/src/main/java/org/sonar/java/JavaFrontend.java

```diff
@@ -0,0 +1,643 @@
+import com.google.common.collect.Multimap;
+import com.sonar.sslr.api.typed.ActionParser;
+import java.io.File;
+import java.util.*;
+
+import org.sonar.java.ast.parser.JavaParser;
+import org.sonar.java.cfg.CFG;
+import org.sonar.java.cfg.CFG.Block;
+import org.sonar.java.model.ExpressionUtils;
+import org.sonar.java.resolve.JavaSymbol;
+import org.sonar.java.resolve.SemanticModel;
+import org.sonar.plugins.java.api.semantic.Symbol;
+import org.sonar.plugins.java.api.semantic.Symbol.MethodSymbol;
+import org.sonar.plugins.java.api.semantic.Type;
+import org.sonar.plugins.java.api.tree.*;
+
+public class JavaFrontend {
+
+  public static class ScannedFile {
+
+    private final CompilationUnitTree tree;
+    private final SemanticModel semantic;
+
+    public ScannedFile(CompilationUnitTree tree, SemanticModel semantic) {
+      this.tree = tree;
+      this.semantic = semantic;
+    }
+
+    public CompilationUnitTree tree() {
+      return tree;
+    }
+
+    public SemanticModel semantic() {
+      return semantic;
+    }
+  }
+
+  private final ActionParser<Tree> parser = JavaParser.createParser();
+
+  public ScannedFile scan(File source, ClassLoader classLoader) {
+    CompilationUnitTree tree = (CompilationUnitTree) parser.parse(source);
+    SemanticModel model = SemanticModel.createFor(tree, classLoader);
+
+    return new ScannedFile(tree, model);
+  }
+
+  public static class TaintSummary {
+
+    private final Set<TaintSource> result;
+    private final Set<DirectTaintSource> methodCalls;
+
+    public TaintSummary(Set<TaintSource> result, Set<DirectTaintSource> methodCalls) {
+      this.result = result;
+      this.methodCalls = methodCalls;
+    }
+
+    public boolean resultCanBeTaintedBy(String s) {
+      for (TaintSource ts: result) {
+        if (ts.toString().equals(s)) {
+          return true;
+        }
+      }
+      return false;
+    }
+
+    public int resultTaintedPaths() {
+      return result.size();
+    }
+
+    public boolean callsMethod(String s) {
+      for (DirectTaintSource ts: methodCalls) {
+        if (ts.toString().equals(s)) {
+          return true;
+        }
+      }
+      return false;
+    }
+
+    public int methodCalls() {
+      return methodCalls.size();
+    }
+
+  }
+
+  public interface TaintSource {
+
+    boolean canBeTainted();
+    TaintSource unionWith(TaintSource other);
+
+  }
+
+  /**
+   * Literals
+   * Non-string operands
+   */
+  public static class TaintFreeSource implements TaintSource {
+
+    @Override
+    public boolean canBeTainted() {
+      return false;
+    }
+
+    @Override
+    public TaintSource unionWith(TaintSource other) {
+      return other;
+    }
+
+    @Override
+    public String toString() {
+      return "taint-free";
+    }
+
+    @Override
+    public int hashCode() {
+      return 0;
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+      if (obj == null || !(obj.getClass().equals(this.getClass()))) {
```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Replace this if-then-else statement by a single return statement. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1126)